src/Estimators/OneBodyDensityMatrices.cpp

```diff
@@ -0,0 +1,1332 @@
+//   int nentries = ng[0] * ng[1] * ng[2];
+// #else
+//   std::vector<int> ng(2);
+//   ng[0]        = basis_size;
+//   ng[1]        = basis_size;
+//   int nentries = ng[0] * ng[1];
+// #endif
+
+//   std::string dname = myName;
+//   hid_t dgid        = H5Gcreate(gid, dname.c_str(), 0);
+
+//   std::string nname = "number_matrix";
+//   hid_t ngid        = H5Gcreate(dgid, nname.c_str(), 0);
+//   for (int s = 0; s < nspecies; ++s)
+//   {
+//     h5desc.emplace_back(species_name[s]);
+//     auto& oh = h5desc.back();
+//     oh.set_dimensions(ng, nindex + s * nentries);
+//     oh.open(ngid);
+//   }
+
+//   if (energy_mat)
+//   {
+//     std::string ename = "energy_matrix";
+//     hid_t egid        = H5Gcreate(dgid, ename.c_str(), 0);
+//     for (int s = 0; s < nspecies; ++s)
+//     {
+//       h5desc.emplace_back(species_name[s]);
+//       auto& oh = h5desc.back();
+//       oh.set_dimensions(ng, eindex + s * nentries);
+//       oh.open(egid);
+//     }
+//   }
+// }
+
+void OneBodyDensityMatrices::warmup_sampling(ParticleSet& pset_target)
+{
+  if (input_.get_sampling() == Samplings::METROPOLIS)
+  {
+    if (!warmed_up)
+    {
+      diffusion(std::sqrt(input_.get_timestep()), rpcur);
+      rpcur += center;
+      if (input_.get_integrator() == Integrators::DENSITY)
+        density_drift(rpcur, rhocur, dpcur, pset_target);
+      else
+        APP_ABORT("OneBodyDensityMatrices::warmup_sampling  invalid integrator");
+    }
+    generate_samples(1.0, pset_target, input_.get_warmup_samples());
+    warmed_up = true;
+  }
+}
+
+
+// OneBodyDensityMatrices::FullPrecReal OneBodyDensityMatrices::evaluate(ParticleSet& P)
+// {
+//   ScopedTimer t(timers[DM_eval]);
+//   if (have_required_traces || !energy_mat)
+//   {
+//     if (check_derivatives)
+//       test_derivatives();
+//     if (input_.get_evaluator() == loop)
+//       evaluate_loop(P);
+//     else if (input_.get_evaluator() == matrix)
+//       evaluate_matrix(P);
+//     else
+//       APP_ABORT("OneBodyDensityMatrices::evaluate  invalid evaluator");
+//   }
+//   return 0.0;
+// }
+
+// void OneBodyDensityMatrices::accumulate(const RefVector<MCPWalker>& walkers,
+//                                         const RefVector<ParticleSet>& psets,
+//                                         const RefVector<TrialWaveFunction>& wfns,
+//                                         RandomGenerator_t& rng)
+// {
+//   ScopedTimer scope_timer(timers_.eval_timer);
+//   for (int iw = 0; iw < walkers.size(); ++iw)
+//   {
+//     if (check_derivatives)
+//       test_derivatives();
+//   if (input_.get_evaluator() == loop)
+//     evaluate_loop(P);
+//   else if (input_.get_evaluator() == matrix)
+//     evaluate_matrix(P);
+//   else
+//     throw logic_error("OneBodyDensityMatrices::accumulate called with invalid evaluator, developer error");
+//   }
+//   return 0.0;
+// }
+
+OneBodyDensityMatrices::FullPrecReal OneBodyDensityMatrices::evaluate_matrix(ParticleSet& P,
```

Comment:
Codecov really interferes with review.  Code is almost unreadable.

Reply:
Punishment for adding code without unit tests. But I think this PR is enough.